### PR TITLE
New witness program definition in BIP141, and related revision in 142 - 144

### PR DESCRIPTION
**Pull request by jl2012**
BIP141: New witness program definition
BIP142: Title change; new address definition
BIP143: Title change and update reference implementation
BIP144: Update reference implementation

### CODE REVIEW

**Comment by CodeShark** [comment]
ACK on the 141 and 143 changes

**Comment by CodeShark** [comment]
ACK on all BIPs.

**Comment by dcousens** [comment]
ACK BIP 142,  **much** improved.

File: bip-0142.mediawiki [hunk 1]
```diff
@@ -9,148 +9,142 @@
...
+However, none of these are as widely adopted as the suboptimal base-58 scriptPubKey template addresses, which is still a standard for the whole eco-system, from wallets, block explorers, merchants, exchanges, to end users. It is believed that the proposed P2WPKH and P2WSH address format is the easiest way for wallets and services to adopt native witness program, which is particularly important in the context of scaling the capacity of the blockchain.
 
-The P2PKH segwit addresses only allow simple payment to a single public key. For arbitrarily complex segwit transactions, the general segwit address is used. The use of z-base-32 eliminates case-sensitiveness and simplifies transformation to and from hexadecimals. The Damm algorithm checksum allows detection of all single-digit errors and all adjacent transposition errors. With the excellent upgradability of witness program, this proposal is also forward compatible to new version witness programs that is not longer than 32 bytes.
+While P2WPKH address is specific for simple payment to a signle public key, P2WSH address allows arbitrarily complex segwit transactions, resemble to the BIP13 P2SH address.
```
Comment (schildbach) [inline bip-0142.mediawiki]: signle > single

**Comment by btcdrak** [comment]
ACK

File: bip-0142.mediawiki (hunk 1 above)
Comment (jl2012) [inline bip-0142.mediawiki]: @schildbach Corrected. Thanks

**Comment by NicolasDorier** [comment]
NACK for 142. This would make a conflict between two versions, breaking client who depended on the base58 to deduce the network of the string.

Main EXT_PUBLIC_KEY 4,136,178,30 <=> TestNet WITNESS_P2WPKH 4

Solution without conflict:
```
base58Prefixes[(int)Base58Type.WITNESS_P2WPKH] = new byte[] { (0x03) };
base58Prefixes[(int)Base58Type.WITNESS_P2WSH] = new byte[] { (40) };

base58Prefixes[(int)Base58Type.WITNESS_P2WPKH] = new byte[] { 0x6 };
base58Prefixes[(int)Base58Type.WITNESS_P2WSH] = new byte[] { (10) };
```

**Comment by jl2012** [comment]
BIP142 revised to resolve the conflict

**Comment by NicolasDorier** [comment]
Thanks, last thing I am thinking. Why making a scheme address specific for v0 and not something general to all segwit payment ? If we use the BIP142 scheme, the problem is that when a new segwit will release, no wallet will be able to send to it... all wallets want to do is sending money, they should not need to implement a new address scheme for a new segwit version.

**Comment by jl2012** [comment]
@NicolasDorier This is actually generalizable, as long as the future version is 20 byte or 32 byte or both. It's very likely that future versions will be 32 byte.

In the very first version of BIP142 I actually made it completely generalizable to any script length but that idea was not very popular: https://github.com/bitcoin/bips/pull/267

**Comment by NicolasDorier** [comment]
well, in future version a new address version should be found. And during a while, all wallet will not be able to send to it.

**Comment by NicolasDorier** [comment]
I see, ACK. In the future a new address scheme will be implemented anyway.

File: README.mediawiki [hunk 2]
```diff
@@ -1,4 +1,4 @@
-People wishing to submit BIPs, first should propose their idea or document to the mailing list. After discussion they should email Luke Dashjr &lt;luke_bipeditor@dashjr.org&gt;. After copy-editing and acceptance, it will be published here.
+﻿People wishing to submit BIPs, first should propose their idea or document to the mailing list. After discussion they should email Luke Dashjr &lt;luke_bipeditor@dashjr.org&gt;. After copy-editing and acceptance, it will be published here.
```
Comment (luke-jr) [inline README.mediawiki]: Don't need a BOM here...

**Pull request merged**